src/test/java/BookstoreRestTest.java

```diff
@@ -0,0 +1,460 @@
+     *   - The Book is no longer available for Sale
+     *   - The Book is from the Bookstore that it was bought from
+     *   - The Bookstore associated with the Sale is the same as the Bookstore that the Book is from
+     * @throws Exception
+     */
+    @Test
+    @Order(6)
+    public void newSale() throws Exception {
+        this.mockMvc.perform(
+                post("/api/newSale")
+                        .param("customerId", this.testCustomerId.toString()))
+                .andExpect(status().isOk());
+
+        //Query repository for newly created test Customer
+        Optional<Customer> optionalCustomer = this.customerRepository.findById(this.testCustomerId);
+        Customer customer = optionalCustomer.get();
+
+        //Get the Customer's ShoppingCart
+        ShoppingCart shoppingCart = customer.getShoppingCart();
+
+        //Test that the ShoppingCart is now empty
+        assert(shoppingCart.getBooks().isEmpty());
+
+        //Get the Sale from the Customer
+        List<Sale> sales = new ArrayList<Sale>(customer.getSales());
+
+        //Test that there is only a single sale
+        assert(sales.size() == 1);
+
+        //Get the Sale from the Customer
+        Sale sale = sales.get(0);
+
+        //Get the Books from the Sale
+        List<Book> books = new ArrayList<Book>(sale.getBooks());
+
+        //Test that the sale only has one book
+        assert(books.size() == 1);
+
+        Book book = books.get(0);
+        //Test that the book is no longer available
+        assert(!book.getAvailable());
+
+        //Test that the Book belongs to TestBookStore
+        assert(book.getBookstore().getId() == this.testBookstoreId);
+
+        //Test that the Sale is associated with TestBookStore
+        ArrayList<Bookstore> bookstores = new ArrayList<Bookstore>(sale.getBookstores());
+        Bookstore bookstore = bookstores.get(0);
+        assert(bookstore.getId() == this.testBookstoreId && bookstore.getName().equals(this.testBookstoreName));
+    }
+
+    /**
+     * Queries for a BookstoreOwner by bookstoreOwnerId via REST, compares JSON returned to expected JSON value
+     * Expected value:{"id":1,"name":"testBookStoreOwner"}
+     * @throws Exception
+     */
+    @Test
+    public void getBookstoreOwner() throws Exception {
+        String expectedResult = String.format("{\"id\":%d,\"name\":\"%s\"}",
+                this.testBookstoreOwnerId,
+                this.testBookstoreOwnerName);
+        this.mockMvc.perform(
+                get("/api/getBookstoreOwner").param("bookstoreOwnerId", this.testBookstoreOwnerId.toString()))
+                .andExpect(status().isOk())
+                .andExpect(content().string(expectedResult));
+    }
+
+    /**
+     * Queries for all BookstoreOwners via REST, compares JSON returned to expected JSON value
+     * Expected value:[{"id":1,"name":"testBookStoreOwner"}]
+     * @throws Exception
+     */
+    @Test
+    public void getBookstoreOwners() throws Exception {
```

Comment:
Curious why the tests above have the @Order annotation but the one's below (including this one) do not. Does Order not matter when testing the one's below, or how does it work?

Reply:
The tests that have @Order specified guarantee the order in which they run, if no annotation is provided, the junit runtime will determine the order, and it is not always the same. In this case, the tests with the @Order annotation need to run in a certain order to ensure the state is correct to simulate a typical scenario involving a Customer purchasing a book from a Bookstore. The order ensures all necessary entities are created. The tests that do not have the @Order annotation are testing the getter endpoints, and there order does not matter.